From: Long Li <longli@microsoft.com>
Date: Mon, 21 Aug 2017 22:31:45 -0700
Patch-mainline: never (code changed in 4.4)
Subject: storvsc: do not assume SG list is continuous when doing bounce buffers
References: bsc#1075410

storvsc checks the SG list for gaps before passing them to Hyper-v
device. If there are gaps, data is copied to a bounce buffer and a
continuous data buffer is passed to Hyper-V.

The check on gaps assumes SG list is continuous, and not chained. This
is not always true. Failing the check may result in incorrect I/O data
passed to the Hyper-v device.

This code path is not used post Linux 4.1.

Signed-off-by: Long Li <longli@microsoft.com>
Acked-by: Martin K. Petersen <martin.petersen@oracle.com>
Acked-by: <ohering@suse.de>

--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -635,17 +635,18 @@ static int do_bounce_buffer(struct scatt
 	for (i = 0; i < sg_count; i++) {
 		if (i == 0) {
 			/* make sure 1st one does not have hole */
-			if (sgl[i].offset + sgl[i].length != PAGE_SIZE)
+			if (sgl->offset + sgl->length != PAGE_SIZE)
 				return i;
 		} else if (i == sg_count - 1) {
 			/* make sure last one does not have hole */
-			if (sgl[i].offset != 0)
+			if (sgl->offset != 0)
 				return i;
 		} else {
 			/* make sure no hole in the middle */
-			if (sgl[i].length != PAGE_SIZE || sgl[i].offset != 0)
+			if (sgl->length != PAGE_SIZE || sgl->offset != 0)
 				return i;
 		}
+		sgl = sg_next(sgl);
 	}
 	return -1;
 }
